drivers/mpu9250/mpu9250.c

```diff
@@ -52,89 +68,223 @@ typedef struct
+  dev->polling_data.acc_x = (rdata & 0x0000FFFF);
   rdata = ioread16(dev->regs + MEM_OFFSET_DATA_ACC_Y);
-  dev->buffer.acc_y = (rdata & 0x0000FFFF);
+  dev->polling_data.acc_y = (rdata & 0x0000FFFF);
   rdata = ioread16(dev->regs + MEM_OFFSET_DATA_ACC_Z);
-  (dev->buffer.acc_z = (rdata & 0x0000FFFF));
+  (dev->polling_data.acc_z = (rdata & 0x0000FFFF));
 
   rdata = ioread16(dev->regs + MEM_OFFSET_DATA_MAG_X);
-  dev->buffer.mag_x = (rdata & 0x0000FFFF);
+  dev->polling_data.mag_x = (rdata & 0x0000FFFF);
   rdata = ioread16(dev->regs + MEM_OFFSET_DATA_MAG_Y);
-  dev->buffer.mag_y = (rdata & 0x0000FFFF);
+  dev->polling_data.mag_y = (rdata & 0x0000FFFF);
   rdata = ioread16(dev->regs + MEM_OFFSET_DATA_MAG_Z);
-  dev->buffer.mag_z = (rdata & 0x0000FFFF);
+  dev->polling_data.mag_z = (rdata & 0x0000FFFF);
+
+  dev->polling_data.timestamp_lo = ioread32(dev->regs + MEM_OFFSET_TIMESTAMP_LOW);
+  dev->polling_data.timestamp_hi = ioread32(dev->regs + MEM_OFFSET_TIMESTAMP_HIGH);
+
+  /* copy data from kernel space buffer into user space */
+  if (count > 0)
+    count = count - copy_to_user(buf, (char *)&dev->polling_data + *offp, count);
+
+  *offp += count;
+
+  return count;
+}
+
+static int read_buffer_data(struct data *dev, char *buf, size_t count, loff_t *offp)
+{
+  int i;
 
-  dev->buffer.timestamp_lo = ioread32(dev->regs + MEM_OFFSET_TIMESTAMP_LOW);
-  dev->buffer.timestamp_hi = ioread32(dev->regs + MEM_OFFSET_TIMESTAMP_HIGH);
+  /* check out of bound access */
+  if ((*offp < 0) || (*offp >= SIZEOF_BUFFER_DATA_T))
+    return 0;
+
+  /* limit number of readable bytes to maximum which is still possible */
+  if ((*offp + count) > SIZEOF_BUFFER_DATA_T)
+    count = SIZEOF_BUFFER_DATA_T - *offp;
+
+  /* Fill structure with dummy data */
+  for (i = 0; i < 1024; i++)
+  {
+    dev->buffer_data.buf_acc_x[i] = 0x1000 | i;
+    dev->buffer_data.buf_acc_y[i] = 0x2000 | i;
+    dev->buffer_data.buf_acc_y[i] = 0x3000 | i;
+
+    dev->buffer_data.buf_gyro_x[i] = 0x4000 | i;
+    dev->buffer_data.buf_gyro_y[i] = 0x5000 | i;
+    dev->buffer_data.buf_gyro_y[i] = 0x6000 | i;
+  }
 
   /* copy data from kernel space buffer into user space */
   if (count > 0)
-    count = count - copy_to_user(buf, (char *)&dev->buffer + *offp, count);
+    count = count - copy_to_user(buf, (char *)&dev->buffer_data + *offp, count);
 
   *offp += count;
 
   return count;
 }
 
+static irqreturn_t irq_handler(int nr, void *data_ptr)
+{
+  struct data *dev = data_ptr;
+  uint32_t irqs;
+  pr_info("Interrupt occured\n");
+
+  /* Determine which interrupt occured */
+  irqs = ioread32(dev->regs + MEM_OFFSET_BUF_ISR);
+
+  if (irqs == 0x1)
+  {
+    pr_info("Received Button[1] interrupt");
+    return IRQ_HANDLED;
+  }
+  else if (irqs == 0x2)
+  {
+    pr_info("Received Button[2] interrupt");
+    return IRQ_HANDLED;
+  }
+  else if (irqs == 0x3)
+  {
+    pr_info("Received Button[1] and Button[2] interrupt");
+    return IRQ_HANDLED;
+  }
+  return IRQ_NONE;
```

Comment:
Since you have to signal the user space process somehow, you need its PID first (using [this](https://stackoverflow.com/a/12434702) seems appropriate).

Do you want to add another `ioctl` command where processes can register for interrupts or something?